src/library_manager/llext_manager.c

```diff
@@ -71,177 +80,219 @@ static int llext_manager_align_unmap(void __sparse_cache *vma, size_t size)
 static int llext_manager_load_data_from_storage(void __sparse_cache *vma, void *s_addr,
 						size_t size, uint32_t flags)
 {
-	int ret = llext_manager_align_map(vma, size, flags);
+	int ret = llext_manager_align_map(vma, size, SYS_MM_MEM_PERM_RW);
 
 	if (ret < 0) {
 		tr_err(&lib_manager_tr, "cannot map %u of %p", size, (__sparse_force void *)vma);
 		return ret;
 	}
 
-	/* TODO: Change attributes for memory to FLAGS  */
-	return memcpy_s((__sparse_force void *)vma, size, s_addr, size);
+	ret = memcpy_s((__sparse_force void *)vma, size, s_addr, size);
+	if (ret < 0)
+		return ret;
+
+	/*
+	 * We don't know what flags we're changing to, maybe the buffer will be
+	 * executable or read-only. Need to write back caches now
+	 */
+	dcache_writeback_region(vma, size);
+
+	ret = llext_manager_update_flags(vma, size, flags);
+	if (!ret && (flags & SYS_MM_MEM_PERM_EXEC))
+		icache_invalidate_region(vma, size);
+
+	return ret;
 }
 
 static int llext_manager_load_module(uint32_t module_id, const struct sof_man_module *mod)
 {
 	struct lib_manager_mod_ctx *ctx = lib_manager_get_mod_ctx(module_id);
 	uint8_t *load_base = (uint8_t *)ctx->base_addr;
+
+	/* Executable code (.text) */
 	void __sparse_cache *va_base_text = (void __sparse_cache *)
-		mod->segment[SOF_MAN_SEGMENT_TEXT].v_base_addr;
-	void *src_txt = (void *)(load_base + mod->segment[SOF_MAN_SEGMENT_TEXT].file_offset);
-	size_t st_text_size = ctx->segment_size[SOF_MAN_SEGMENT_TEXT];
+		ctx->segment[LIB_MANAGER_TEXT].addr;
+	void *src_txt = (void *)(load_base + ctx->segment[LIB_MANAGER_TEXT].file_offset);
+	size_t text_size = ctx->segment[LIB_MANAGER_TEXT].size;
+
+	/* Read-only data (.rodata and others) */
 	void __sparse_cache *va_base_rodata = (void __sparse_cache *)
-		mod->segment[SOF_MAN_SEGMENT_RODATA].v_base_addr;
+		ctx->segment[LIB_MANAGER_RODATA].addr;
 	void *src_rodata = (void *)(load_base +
-				    mod->segment[SOF_MAN_SEGMENT_RODATA].file_offset);
-	size_t st_rodata_size = ctx->segment_size[SOF_MAN_SEGMENT_RODATA];
+				    ctx->segment[LIB_MANAGER_RODATA].file_offset);
+	size_t rodata_size = ctx->segment[LIB_MANAGER_RODATA].size;
+
+	/* Writable data (.data, .bss and others) */
+	void __sparse_cache *va_base_data = (void __sparse_cache *)
+		ctx->segment[LIB_MANAGER_DATA].addr;
+	void *src_data = (void *)(load_base +
+				  ctx->segment[LIB_MANAGER_DATA].file_offset);
+	size_t data_size = ctx->segment[LIB_MANAGER_DATA].size;
+
+	/* .bss, should be within writable data above */
+	void __sparse_cache *bss_addr = (void __sparse_cache *)
+		ctx->segment[LIB_MANAGER_BSS].addr;
+	size_t bss_size = ctx->segment[LIB_MANAGER_BSS].size;
 	int ret;
 
+	/* Check, that .bss is within .data */
+	if (bss_size &&
+	    ((uintptr_t)bss_addr + bss_size <= (uintptr_t)va_base_data ||
+	     (uintptr_t)bss_addr >= (uintptr_t)va_base_data + data_size)) {
+		if ((uintptr_t)bss_addr + bss_size == (uintptr_t)va_base_data &&
+		    !((uintptr_t)bss_addr & (PAGE_SZ - 1))) {
```

Comment:
ALIGN_DOWN

Reply:
So you can use `IS_ALIGNED`

Reply:
I could, but I don't want to use division where masking works too.